src/CityOnPlanet.cpp

```diff
@@ -199,64 +284,106 @@ CityOnPlanet::~CityOnPlanet()
+			return (true);
+	}
+	return (false);
+}
+
+void CityOnPlanet::setCityCellsOccupied(uint32_t x, uint32_t z, uint32_t layout)
+{
+	union bits {
+		uint16_t whole;
+		uint8_t bytes[2];
+	} bits;
+
+	bits.whole = ((uint16_t)(0xff00 << (7 - layout))) >> (z & 7);
+
+	for (uint32_t i = 0; i <= layout; i++) {
+#ifdef __BIG_ENDIAN__
+		citybits[x + i][z >> 3] |= bits.bytes[0];
+		citybits[x + i][(z >> 3) + 1] |= bits.bytes[1];
+#else
+		// x86/x64
+		citybits[x + i][z >> 3] |= bits.bytes[1];
+		citybits[x + i][(z >> 3) + 1] |= bits.bytes[0];
+#endif
+	}
+}
+
 CityOnPlanet::CityOnPlanet(Planet *planet, SpaceStation *station, const Uint32 seed)
 {
 	// beware, these are not used in this function, but are used in subroutines!
 	m_planet = planet;
 	m_frame = planet->GetFrame();
 	m_detailLevel = Pi::detail.cities;
 
-	m_buildings.clear();
-	m_buildings.reserve(DEFAULT_NUM_BUILDINGS);
-
+	// Group outcalls together to make things easier for cpu cache
 	const Aabb &aabb = station->GetAabb();
 	const matrix4x4d &m = station->GetOrient();
-	const vector3d p = station->GetPosition();
-
-	const vector3d mx = m * vector3d(1, 0, 0);
-	const vector3d mz = m * vector3d(0, 0, 1);
+	const vector3d station_position = station->GetPosition();
+	const std::string &stationmodelname = station->GetStationType()->ModelName();
+	const double bodyradius = planet->GetSystemBody()->GetRadius();			  // cache for bodyradius value
+	const double pop = 1000000000 * planet->GetSystemBody()->GetPopulation(); // GetPopulation() returns population in whole billions, as in 0.5 -> 500M
+	const int atmo = planet->GetSystemBody()->HasAtmosphere() ? config_atmo_size : config_airless_size;
 
 	Random rand;
 	rand.seed(seed);
 
-	int population = planet->GetSystemBody()->GetPopulation();
-	int cityradius;
+	const vector3d mx = m * vector3d(1, 0, 0);
+	const vector3d mz = m * vector3d(0, 0, 1);
+	citybuildinglist_t *buildings = &s_buildingList;
 
-	if (planet->GetSystemBody()->HasAtmosphere()) {
-		population *= 1000;
-		cityradius = (population < 200) ? 200 : ((population > START_SEG_SIZE) ? START_SEG_SIZE : population);
-	} else {
-		population *= 100;
-		cityradius = (population < 250) ? 250 : ((population > START_SEG_SIZE_NO_ATMO) ? START_SEG_SIZE_NO_ATMO : population);
-	}
+	const int cellsize_i = 50;
+	const double cellsize = double(cellsize_i);
+	int cityradius = 0;
 
-	citybuildinglist_t *buildings = &s_buildingList;
-	vector3d cent = p;
-	const int cellsize_i = 80;
-	const double cellsize = double(cellsize_i);						// current widest building = 92
-	const double bodyradius = planet->GetSystemBody()->GetRadius(); // cache for bodyradius value
+	for (int i = 0; i < 5; i++) {
+		if (pop > population2radius[i]) {
+			cityradius = (atmo * cityradius_mod[i]) + rand.Int32(cityradius_rnd[i]);
+			break;
+		}
+	}
 
-	static const int gmid = (cityradius / cellsize_i);
-	static const int gsize = gmid * 2;
+	if (cityradius > 98)
+		cityradius = 98;
+	if (cityradius < 10)
+		cityradius = 10;
```

Comment:
This needs a comment to explain the parameter space and why these constants were chosen.